File: TRPO/lab6/myproject/bin/main/com/example/factory/UserFactory.kt
package com.example.factory

import com.example.model.UserType 
import com.example.model.IntegerType
import com.example.model.GPSCoordinateType

class UserFactory {
    private val types = listOf(IntegerType(), GPSCoordinateType())

    fun getTypeNames() = types.map { it.typeName() }

    fun getBuilderByName(name: String): UserType {
        return types.find { it.typeName() == name } 
            ?: throw IllegalArgumentException("Тип $name не поддерживается")
    }
}

File: TRPO/lab5/scala_lab5/src/main/scala/UserFactory.scala
class UserFactory {
  private val types = List(new IntegerType, new GPSCoordinateType)

  def getTypeNameList: List[String] = types.map(_.typeName)
  
  def getBuilderByName(name: String): UserType = 
    types.find(_.typeName == name).getOrElse(throw new IllegalArgumentException)
}